piet-coregraphics/src/text.rs

```diff
@@ -0,0 +1,433 @@
+            .take_while(|b| match b {
+                b' ' | b'\t' | b'\n' | b'\r' => true,
+                _ => false,
+            })
+            .count();
+        let height = typo_bounds.ascent + typo_bounds.descent + typo_bounds.leading;
+        // this may not be exactly right, but i'm also not sure we ever use this?
+        //  see https://stackoverflow.com/questions/5511830/how-does-line-spacing-work-in-core-text-and-why-is-it-different-from-nslayoutm
+        let cumulative_height =
+            (self.line_y_positions[line_number] + typo_bounds.descent + typo_bounds.leading).ceil();
+        Some(LineMetric {
+            start_offset,
+            end_offset,
+            trailing_whitespace,
+            baseline: typo_bounds.ascent,
+            height,
+            cumulative_height,
+        })
+    }
+
+    fn line_count(&self) -> usize {
+        self.line_y_positions.len()
+    }
+
+    // given a point on the screen, return an offset in the text, basically
+    fn hit_test_point(&self, point: Point) -> HitTestPoint {
+        let mut line_num = self
+            .line_y_positions
+            .iter()
+            .position(|y| y >= &point.y)
+            // if we're past the last line, use the last line
+            .unwrap_or_else(|| self.line_y_positions.len().saturating_sub(1));
+        // because y_positions is the position of the baseline, check that we don't
+        // fall between the preceding baseline and that line's descent
+        if line_num > 0 {
+            let prev_line = self.unwrap_frame().get_line(line_num - 1).unwrap();
+            let typo_bounds = Line::new(&&prev_line).get_typographic_bounds();
+            if self.line_y_positions[line_num - 1] + typo_bounds.descent >= point.y {
+                line_num -= 1;
+            }
+        }
+        let line: Line = self
+            .unwrap_frame()
+            .get_line(line_num)
+            .map(Into::into)
+            .unwrap();
+        let fake_y = self.line_y_positions[line_num];
+        // map that back into our inverted coordinate space
+        let fake_y = -(self.frame_size.height - fake_y);
+        let point_in_string_space = CGPoint::new(point.x, fake_y);
+        let offset_utf16 = line.get_string_index_for_position(point_in_string_space);
+        let offset = match offset_utf16 {
+            // this is 'kCFNotFound'.
+            // if nothing is found just go end of string? should this be len - 1? do we have an
+            // implicit newline at end of file? so many mysteries
+            -1 => self.string.len(),
+            n if n >= 0 => {
+                let utf16_range = line.get_string_range();
+                let utf8_range = self.line_range(line_num).unwrap();
+                let line_txt = self.line_text(line_num).unwrap();
+                let rel_offset = (n - utf16_range.location) as usize;
+                let mut off16 = 0;
+                let mut off8 = 0;
+                for c in line_txt.chars() {
```

Comment:
I'd recommend calling into a helper function; https://github.com/linebender/piet/blob/master/piet-direct2d/src/text.rs#L230 is reasonable though we should have better ways of avoiding cross-crate duplication.

Reply:
opened #183 to track sharing code between crates.